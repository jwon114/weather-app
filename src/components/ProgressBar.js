import React from 'react';

export default class ProgressBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      timer: props.secondsCountdown
    };
  }

  componentDidMount() {
    if (this.props.startTimer) {
      this.timerId = setInterval(() => this.setState((prevState) => ({
        timer: prevState.timer--
      })), 1000);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.timer < 0) {
      clearInterval(this.timerId);
      this.setState(() => ({ timer: 0 }));
      this.props.handleTimerFinished();
    }
  }

  render() {
    const percentage = 100 - ((this.state.timer / this.props.secondsCountdown) * 100)

    return (
      <div className="progress-bar">
        <div className="progress-bar__text">{`Reloading in ${this.state.timer}s`}</div>
        <div className="progress-bar__container">
          <div className="progress-bar__filler" style={{ width: `${percentage}%`}}></div>
        </div>
      </div>
    );
  }
}