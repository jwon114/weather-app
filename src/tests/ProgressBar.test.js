import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import { render } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import ProgressBar from '../components/ProgressBar';

test('should render ProgressBar correctly', () => {
  const renderer = new ReactShallowRenderer();
  renderer.render(<ProgressBar />);
  
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

test('should display 60 second countdown text', () => {
  const { getByText } = render(<ProgressBar 
    startTimer={true} 
    initialTime={60}
    handleTimerFinished={() => {}} />);

  expect(getByText(/reloading in 60s/i)).toBeInTheDocument();
});

test('sets ProgressBar timer state to initial time when mounted', () => {
  const renderer = TestRenderer.create(<ProgressBar 
    startTimer={false} 
    initialTime={60}
    handleTimerFinished={() => {}} />);

  const app = renderer.getInstance();
  
  expect(app.state.timer).toEqual(60);
});

test('sets ProgressBar timer state to 0 when timer finishes and calls handleTimerFinished function', () => {
  jest.useFakeTimers();
  let renderer = TestRenderer.create(<ProgressBar 
    startTimer={true} 
    initialTime={60}
    handleTimerFinished={() => {}} />);

  const app = renderer.getInstance();
  expect(app.state.timer).toEqual(60);
  expect(setInterval).toHaveBeenCalled();
  expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);  
  jest.advanceTimersByTime(60000);
  expect(app.state.timer).toEqual(0);
  expect(app.props.handleTimerFinished).toHaveBeenCalled();
});

// test('starts ProgressBar timer when startTimer prop changes', () => {
//   jest.useFakeTimers();
//   let renderer = TestRenderer.create(<ProgressBar 
//     startTimer={false} 
//     initialTime={60}
//     handleTimerFinished={() => {}} />);

//   const app = renderer.getInstance();
//   expect(app.state.timer).toEqual(60);
//   app.update();
//   expect(setInterval).toHaveBeenCalled();
//   expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);  
//   jest.advanceTimersByTime(60000);
//   expect(app.state.timer).toEqual(0);
//   expect(app.props.handleTimerFinished).toHaveBeenCalled();
// });

