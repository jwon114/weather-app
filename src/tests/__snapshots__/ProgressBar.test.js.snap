// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProgressBar correctly 1`] = `
<div
  className="progress-bar"
>
  <span
    className="progress-bar__text"
  >
    Reloading in 60s
  </span>
  <div
    className="progress-bar__container"
  >
    <div
      className="progress-bar__filler"
      style={
        Object {
          "width": "0%",
        }
      }
    />
  </div>
</div>
`;
